Add PubSub dead letter topic support

Taction opened the pull request:
Signed-off-by: zhangchao <zchao9100@gmail.com>

# Description

<!--
Please explain the changes you've made.
-->
Add PubSub's DeadLetter Topic support.

## Issue reference

<!--
We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.
-->

Please reference the issue this PR will close: #2217

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

yaron2 commented on pkg/runtime/runtime.go:
```diff
@@ -94,6 +94,7 @@ const (
 	bindingsConcurrencyParallel   = "parallel"
 	bindingsConcurrencySequential = "sequential"
 	pubsubName                    = "pubsubName"
+	deadLetterKeyFormat           = "%s||%s" // componentName||topicName
```
You might have two different components with the same name, deployed to different namespaces:

mypubsub deployed to ns1, and mypubsub deployed to ns2.
Does this need to be changed to include a namespace?

Taction replied:
Yes, I'll fix it.

Taction replied:
After a view of the code, components in the different namespaces will be ignored. https://github.com/dapr/dapr/blob/master/pkg/runtime/runtime.go#L1754

yaron2 commented:
@Taction please resolve conflicts. This is a great PR and want to get this into 1.8!

daixiang0 commented:
@Taction is it ready for review as I see WIP in the title?

Taction commented:
@daixiang0 It is ready for review. I am adding unit and e2e tests for it.

yaron2 commented on pkg/runtime/pubsub/subscription.go:
```diff
@@ -1,11 +1,12 @@
-	Scopes     []string          `json:"scopes"`
+	PubsubName      string            `json:"pubsubname"`
+	Topic           string            `json:"topic"`
+	DeadLetterTopic string            `json:"dead_letter_topic"`
```
Should be `deadLetterTopic` to be consistent with the subscription CRD definition.

Taction replied:
Done

yaron2 commented on pkg/runtime/pubsub/subscriptions.go:
```diff
@@ -41,11 +41,12 @@ const (
-		Routes     RoutesJSON        `json:"routes"` // Multiple routes from v2alpha1
+		PubsubName      string            `json:"pubsubname"`
+		Topic           string            `json:"topic"`
+		DeadLetterTopic string            `json:"dead_letter_topic"`
```
Should be `deadLetterTopic` to be consistent with the subscription CRD definition.

Taction replied:
Done

yaron2 commented:
@Taction I reviewed, overall this looks good! I put two small comments.

There's one thing we need to ensure before accepting this: that resiliency policies can be used to determine max retries before going to DLQ.

Taction commented:
@yaron2 Thanks for your review. Fixed.
Resiliency policies can be used to determine max retries before going to DLQ, and I have added a unit test case to ensure that.

yaron2 commented:
@Taction this PR looks good! I'm ready to merge as soon as an e2e test is added.

Taction commented:
Thanks for your words, I have added it to subscribing app's info. The e2e is in progress, if there is still something wrong I'll look into them tomorrow.

Taction commented:
@yaron2 e2e test passed.

yaron2 commented on pkg/runtime/runtime.go:
```diff
@@ -588,39 +611,44 @@ func (a *DaprRuntime) beginPubSub(name string, ps pubsub.PubSub) error {
 					log.Errorf("error serializing cloud event in pubsub %s and topic %s: %s", name, msg.Topic, err)
-					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Retry)), msg.Topic, 0)
-					return err
+					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
```
See comment here: https://github.com/dapr/dapr/pull/4399/files#r861881846.

yaron2 commented on pkg/runtime/runtime.go:
```diff
@@ -588,39 +611,44 @@ func (a *DaprRuntime) beginPubSub(name string, ps pubsub.PubSub) error {
 					log.Errorf("error deserializing cloud event in pubsub %s and topic %s: %s", name, msg.Topic, err)
-					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Retry)), msg.Topic, 0)
-					return err
+					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
```
See comment here: https://github.com/dapr/dapr/pull/4399/files#r861881846.

yaron2 commented:
@Taction I found one issue across multiple places that needs a small fix in the retry/drop logic. Other than that, LGTM.

yaron2 reviewed (approved):
LGTM.

Great work here @Taction 👍

XavierGeerinck commented:
❤️ was missing this feature!!!

yaron2 commented:
@Taction there are inconsistent failures with the resiliency + dead letter unit tests. See an example here: https://github.com/dapr/dapr/runs/6284493034?check_suite_focus=true.

Can you please look into why the test is flaky?

Taction commented:
@yaron2 Sure, I'll look into this.

amulyavarote commented:
> @yaron2 Sure, I'll look into this.

Thank you so much!
